### PR TITLE
add color picker

### PR DESCRIPTION
**Pull request by andrewdang17**

### CODE REVIEW

**Comment by andrewdang17** [comment]
@jamesmk just looking to get this reviewed. I don't think we talked about styling the color picker and also it looks like we'll want a GIF for the documentation.

File: app/assets/javascripts/fae/form/inputs/_color.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        var colors = this.color.colors;
+
+        // default to HEX if Alpha is 1
+        if ($elm.val() && colors.alpha > 0.9) {
+          $elm.val('#' + colors.HEX);
+        }
+      }
+    }
+
+    $('.color').colorPicker(settings);
```
Comment (jamesmk) [inline app/assets/javascripts/fae/form/inputs/_color.js]: Can you append "js" and use something a little more specific, like `js-color-picker`?

Comment (andrewdang17) [reply app/assets/javascripts/fae/form/inputs/_color.js]: updated and now on stage

**Comment by jamesmk** [comment]
@andrewdang17 Can you add the color picker to the release form? That form is meant to showcase all the available form helpers.

In Chrome, I'm not able to enter a custom hex or delete the hex altogether. If I double click to select text, it automatically puts the cursors at the end. When I delete the text and blur the field, "#FFFFFF" get's automatically selected.

File: app/assets/stylesheets/fae/base.scss [hunk 2]
```diff
@@ -67,6 +67,7 @@
   'modules/forms/asset-actions',
   'modules/forms/select',
   'modules/forms/date',
+  'modules/forms/colorpicker',
```
Comment (tshedor) [inline app/assets/stylesheets/fae/base.scss]: My two cents but this looks like a stylesheet provided by the library. The styles in this folder are meant to override the default stylesheet provided by a library (the best example is how `simplemde.min.css` and `_simple-mde.scss` work together; the worst example is `_select.scss`). That way, when an upgrade is necessary, the CSS can be quickly updated and the modifications to integrate with Fae's appearance remain untouched.

If this isn't a library-provided CSS file (it's hard to tell from tinyColorPicker's repo/demo), please 
- avoid [over-qualifying elements as selectors](https://github.com/brigade/scss-lint/blob/master/lib/scss_lint/linter/README.md#qualifyingelement)
- use [Fae's variables for grays](https://github.com/wearefine/fae/blob/master/app/assets/stylesheets/fae/globals/imports/_variables.scss#L47-L55)
- if you have to use an `!important`, make a note why

File: app/assets/stylesheets/fae/modules/forms/_colorpicker.scss [hunk 3]
```diff
@@ -1,9 +1,9 @@
-input.string.js-color-picker {
-  width: 180px;
+.js-color-picker {
+  width: 180px !important; // overrides input.string input
 }
 
 div.cp-color-picker {
```
Comment (andrewdang17) [inline app/assets/stylesheets/fae/modules/forms/_colorpicker.scss]: this 'over-qualifying selector' is unavoidable

**Comment by andrewdang17** [comment]
@jamesmk all the changes are updated on stage now

File: app/assets/stylesheets/fae/modules/forms/_colorpicker.scss (hunk 3 above)
Comment (tshedor) [reply app/assets/stylesheets/fae/modules/forms/_colorpicker.scss]: FWIW you can almost always get around `!important` in these situations with the specificity hack of doubling a selector. Here it'd look like `.js-color-picker.js-color-picker`, but that can be compared against the FINE standards of the moment (cc @aaronransley ). I'd personally favor doubling over an `!important`.

File: app/assets/stylesheets/fae/modules/forms/_colorpicker.scss [hunk 4]
```diff
@@ -0,0 +1,57 @@
+.js-color-picker {
```
Comment (tshedor) [inline app/assets/stylesheets/fae/modules/forms/_colorpicker.scss]: The purpose of `.js-` prefixed classes is [that they never interact with CSS](http://nicolasgallagher.com/about-html-semantics-front-end-architecture#javascript-specific-classes). This should be a regular class instead (i.e. `.input-color-picker` or simply `.color-picker`) and that class name should be included when you're defining `input_class` on `fae_color_picker`

File: app/assets/stylesheets/fae/modules/forms/_colorpicker.scss [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  width: 180px !important; // overrides input.string input
+}
+
+div.cp-color-picker {
+  border: 1px solid $c-darker-grey;
+  padding: 10px 10px 0;
+  background: #eee;
```
Comment (tshedor) [inline app/assets/stylesheets/fae/modules/forms/_colorpicker.scss]: Should be `$c-lightest-grey`

File: app/assets/stylesheets/fae/modules/forms/_colorpicker.scss [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    border: 1px solid $c-darker-grey;
+  }
+
+  .cp-z-cursor {
+    border-width: 5px;
+    margin-top: -5px;
+  }
+
+  .cp-alpha-cursor {
+    background: #eee;
```
Comment (tshedor) [inline app/assets/stylesheets/fae/modules/forms/_colorpicker.scss]: Should be `$c-lightest-grey`

File: app/assets/stylesheets/fae/modules/forms/_colorpicker.scss [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  border-radius: 3px;
+
+  .cp-alpha {
+    margin: 10px 0;
+    height: 6px;
+    border-radius: 6px;
+    overflow: visible;
+    border: 1px solid $c-darker-grey;
+    box-sizing: border-box;
+    background: linear-gradient(to right, rgba(238, 238, 238, 1) 0%, rgba(238, 238, 238, 0) 100%);
```
Comment (tshedor) [inline app/assets/stylesheets/fae/modules/forms/_colorpicker.scss]: Scss can convert hex to RGB naturally, so these should also use variables: `rgba($c-lightest-grey, 1)` and `rgba($c-lightest-grey, 0)`

File: app/assets/stylesheets/fae/modules/forms/_colorpicker.scss (hunk 3 above)
Comment (aaronransley) [reply app/assets/stylesheets/fae/modules/forms/_colorpicker.scss]: +1 on the double selector hack. It's not pretty but beats an `!important` if possible. Thanks for the work on this feature @andrewdang17!

File: app/assets/javascripts/fae/form/inputs/_color.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        var colors = this.color.colors;
+
+        // default to HEX if Alpha is 1
+        if ($elm.val() && colors.alpha > 0.9) {
+          $elm.val('#' + colors.HEX);
+        }
+      }
+    }
+
+    $('.color-picker').colorPicker(settings);
```
Comment (tshedor) [inline app/assets/javascripts/fae/form/inputs/_color.js]: This needs to be a `.js-` prefixed class. The CSS should stylize the unprefixed class, and the JS prefix should manage whatever JS targeting. The element has **both** `js-color-picker color-picker` classes. The extra `color-picker` in `form_helper.rb` changes to `js-color-picker`.

[Previous comment, noting the inverse](https://github.com/wearefine/fae/pull/345/files/274dc0e927e7056cdb718d373047022c5cfe6152#r152445430):
> The purpose of .js- prefixed classes is that they never interact with CSS. This should be a regular class instead (i.e. .input-color-picker or simply .color-picker) and that class name should be included when you're defining input_class on fae_color_picker

**Pull request merged**